resolves #648 refresh documentation (#649)
asciidoctor/asciidoctor-browser-extension: docs/modules/ROOT/pages/features.adoc

> = Features

Reviewer: I did a few screenshots for the Chrome WebStore page:

<img src="https://lh3.googleusercontent.com/am_TCcxS0oJiRQ909Tg8ajwWFoRGJPjPwH5S4LfQr4tK_mymFmNi9aa2ThH_PN40Bk0Q0hRf4UHTL6ExRLO_zuTjDw=w640-h400-e365-rj-sc0x00ffffff"/>

We might want to use them. I can also take another of screenshots.

Author: Reusing the screen captures from the others makes for a consistent experience.

I did like the animated gif on the vs-code extension:  https://github.com/asciidoctor/asciidoctor-vscode#readme

Author: Let's use what you've taken before, for consistency, or I'm also happy to postpone the decision until the dust settles :-)

[#set-asciidoc-attributes]
== Use AsciiDoc attributes

After the extension applies its own Built-in attributes to document, you can set AsciiDoc attributes that affect the preview in the following order by the:

* extension option xref:custom-attributes.adoc[Custom attributes string]
* <<url-query-parameters>>
* document including any attribute settings included from other documents

This makes Extension very adaptable for your every day, fast and effective use of AsciiDoc.

[#url-query-parameters]
== URL query parameters

You may provide attributes settings to the extension via the URL that you use to view the AsciiDoc.

*Use one parameter*

For example: To condense a table of contents to one level its current setting, add `?toclevels=1` to the end of the URL

.URL query parameters
[,url,subs=+quotes]
----
https://example.com/page.adoc**?toclevels=1**
----

*Add a second parameter*

To add a second parameter, add an ampersand (`&`) to the end of the last, followed by the next attribute setting.

For example: To send a first and last name to the document via the URL, you might add `?first=Irma&last=Writer` to the end of the URL:

.URL query parameters
[,url,subs=+quotes]
----

> https://example.com/page.adoc**?first=Irma&last=Write**

Reviewer: ```suggestion
https://example.com/page.adoc**?first=Irma&last=Writer**
```

This may save you time:

* Writing simple notes
* Converting files from markdown to AsciiDoc gradually.

Learn more about these two text markup formats at:

* xref:asciidoc::asciidoc-vs-markdown.adoc[Compare AsciiDoc to Markdown] via AsciiDoc (the language)
* xref:asciidoctor:migrate:markdown.adoc[Migrate from Markdown to Asciidoctor] via Asciidoctor (the generator)

[#try-it-yourself]
== Try it yourself

To reproduce the example shown in <<example>>, do the following:

:exfile: ex-writing-zen.adoc

. Install the {project-name} using the xref:install.adoc[] page
. Create `{exfile}` from the text below or xref:attachment${exfile}[download] it.
+
.ex-writing-zen.adoc
-----
include::attachment${exfile}[]
-----

. Open `{exfile}` in your browser. You may either:

* Open it by pressing kbd:[Control,O]
* Drag and drop `{exfile}` onto your browser
* Double-click `{exfile}`
+
NOTE: For double-click to work, you may need to change your computer settings to associate `.adoc` with your web browser.

. Try different features with the following:

.. Toggle the {project-name} activation by clicking its extension icon in the browser bar.
+
You'll see the AsciiDoc behind the rendered HTML page.

> .. Change the xref:options.adoc#theme-stylesheet[Theme/Stylesheet] and reload.

Reviewer: ```suggestion
.. Change the xref:theme-stylesheet.adoc[] and reload.
```

.URL query parameters
[,url,subs=+quotes]
----
file:///C:/Users/User/Documents/ex-query.adoc**?first=Irma&last=Writer**
----

You should see the following result:

[indent=0]
====
[discrete]
= Hello Irma Writer!

Query parameters example
====

== Change Themes and Stylesheets

Change the stylesheet, CSS, or both, that are used to render HTML5 from the AsciiDoc.
For example, you may prefer a darker or a more compact presentation style.

The Theme/Stylesheet can be changed for:

* all files previewed via the extension xref:theme-stylesheet-option.adoc[]

* the current file with <<url-query-parameters>>.

== View AsciiDoc files locally and remote

View AsciiDoc files from different locations.

The extension allows you to view AsciiDoc files from your local file system and from the Web:

* Local URLs start with `file://`
* Remote URLs start with `https://` (this includes developer platforms such as GitHub or GitLab)
+
TIP: Use _Raw Mode_ to view AsciiDoc files using the {project-name} and not native rendering provided by developer platforms.

NOTE: Your settings are applied on both local and remote files.

* xref:options.adoc[]
* <<url-query-parameters>>

> == Exploit diagrams as text

Reviewer: It might be better to use "Make use of" or "Leverage" than "Exploit"? I feel like "Exploit" is a bit pejorative no?

```suggestion
== Make use of diagrams as text
```

= Features
:url-asciidoctor-sc-image: https://cdn.jsdelivr.net/gh/asciidoctor/asciidoctor/screenshot.png
:source-language: asciidoc

The main feature is viewing AsciiDoc as HTML5 directly in the browser.

> Other features cover where those AsciiDoc files are located, and how they are rendered.

Reviewer: I don't know what you mean by "where those AsciiDoc files are located"?

I would say something like: "Other features include automatic reload, built-in and custom themes, AsciiDoc attributes as query parameters, diagrams using Kroki and more."

You can exercise these features in the  <<try-it-yourself>> section of this page.

[#example]
== View AsciiDoc as HTML5

The screen capture below shows raw AsciiDoc on the left, and rendered HTML5 on the right.

If the extension is inactive, you'll see the raw AsciiDoc.
If it is active, you'll see the rendered HTML5.

Reproduce this example yourself by following the instructions further below in <<try-it-yourself>>.

image::{url-asciidoctor-sc-image}["Screen capture of two screens, showing raw AsciiDoc and rendered as HTML5"]

== Auto reload

> If the content of the page is changed, the rendered view is automatically refreshed.

Reviewer: That's partially true, the extension will periodically poll for changes and reload/refresh the preview if the content changed.

This means that if you edit the viewed AsciiDoc and save, the rendered view will change.

See xref:options.adoc#poll-for-changes[Poll for Changes] on the options page.

[[set-asciidoc-attributes]]
== Use AsciiDoc attributes

You can set AsciiDoc attributes in the {project-name} to control how your AsciiDoc document is rendered to HTML, without affecting the original document.

The final view is determined by attributes and settings in the:

* Document, including any attributes included from other documents
* Extension option xref:custom-attributes-option.adoc[Custom attributes string]
* <<url-query-parameters>> to the {project-name}

> The attribute settings are applied in the above order of precedence, with Query paramaters being applied last.

Reviewer: ```suggestion
The attribute settings are applied in the above order of precedence, with query parameters being applied last.
```

[#url-query-parameters]
== URL query parameters

You may use URL query parameters to change how the extension previews AsciiDoc. You can set:

> * attribute values with `?attribute=<value>`

Reviewer: ```suggestion
* attribute values with `?<attribute-name>=<value>`
```

* stylesheet used with `?stylesheet=<stylesheet name>` 

*Where to put the query parameters*

Query parameters are indicated by the first (`?`) in the URL and are terminated by the character (`#`) or the end of the URL.

> TIP: If setting attributes or the stylesheet via the URL is not working, check that the Query Parameters are before any (`#`) character in the URL.

Reviewer: ```suggestion
TIP: If setting attributes or the stylesheet via the URL is not working, check that the query parameters are before any (`#`) character in the URL.
```

*Add the first query parameter (`?`)*

> For example: To condense a table of contents to one level its current setting, add `?toclevels=1` to URL, after the document name, and before any (`#`) character.

Reviewer: ```suggestion
For example, to reduce the depth of section levels that are displayed in the table of contents to one, add `?toclevels=1` to the URL (after the document name, and before any `#` character):
```

.URL query parameters
[,url,subs=+quotes]
----
https://example.com/page.adoc**?toclevels=1**
----

> *Add a second and more parameters (`&`)*

Reviewer: It should be "or" no?

```suggestion
*Add a second or more parameters (`&`)*
```

To add one more parameter, add an ampersand (`&`) to the end of the last query parameter, followed by the next attribute setting.

It will look like this:

URL query parameters
[,url,subs=+quotes]
----
https://example.com/page.adoc**?toclevels=1&stylesheet=maker**
----
*Example*

You have a document `ex-query.adoc` that uses two attributes `\{first}` and `\{last}` to say "Hello \{first} \{last}!" in the document title.

.ex-query.adoc
[subs="-attributes",indent=0]
----
= Hello {first} {last}!

Query parameters example
----

TIP: Download the above document from xref:attachment$ex-query.adoc[ex-query.adoc]

> Then, to send a first and last name to the document via the URL, you would add `?first=Irma&last=Writer` to the URL directly after the `ex-query.adoc` and before any (`#`) character.

Reviewer: ```suggestion
Then, to send a first and last name to the document via the URL, you would add `?first=Irma&last=Writer` to the URL directly after the `ex-query.adoc` and before any `#` character.
```

.URL query parameters
[,url,subs=+quotes]
----
file:///C:/Users/User/Documents/ex-query.adoc**?first=Irma&last=Writer**
----

You should see the following result:

[indent=0]
====
[discrete]
= Hello Irma Writer!

Query parameters example
====

== Change Themes and Stylesheets

Change the stylesheet, CSS, or both, that are used to render HTML5 from the AsciiDoc.
For example, you may prefer a darker or a more compact presentation style.

The Theme/Stylesheet can be changed for:

* all files previewed via the extension xref:theme-stylesheet-option.adoc[]

* the current file with <<url-query-parameters>>.

> == View AsciiDocs locally and remote

Reviewer: ```suggestion
== View AsciiDoc files locally and remote
```

View AsciiDoc files from different locations.

The extension allows you to view AsciiDoc files from your local file system and from the web:

* Local URLs start with `file://`

> * Remote URLs start with `https://`
> +
> This includes hosted Git platforms

Reviewer: ```suggestion
* Remote URLs start with `https://` (this includes hosted Git platforms)
```

+

> TIP: *Raw Mode*: To view AsciiDoc files using the {project-name} and not the native AsciiDoc rendering if offered, use the provider's *raw* viewing mode for the file.

Reviewer: I think we should explicitly mention GitLab and GitHub. We can even add a link to: https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#viewing-or-copying-the-raw-file-content

GitLab does not have a documentation page on this feature :(

```suggestion
TIP: Use _Raw Mode_ to view AsciiDoc files using the {project-name} and not native rendering provided by developer platforms such as GitHub or GitLab.
```

Local or remote, your settings in the following are applied:

* xref:options.adoc[]
* <<url-query-parameters>>

== Exploit diagrams as text

> AsciiDoc already allows you to write expressive books, articles, text and documentation and if documenting code to keep it next to your code.

Reviewer: I don't understand this sentence, especially after the "and if..."

Author: AsciiDoc already allows you to write expressive books, articles, text and documentation, and when documenting code to keep it next to your code.

Reviewer: ```suggestion
Using diagrams as text within AsciiDoc allows you to create visual diagrams using plain text, providing a cohesive and unified approach to content creation.
```

.Three step process diagram
image::three-step-process.svg[]

With diagrams as text, you can do the same with diagrams. 

> There are many types of diagrams that can be generated this way from the very simple, to the data-science style of diagrams.

Reviewer: ```suggestion
There are many types of diagrams that can be generated this way from basic block diagrams to complex technical architecture diagrams including data-science visualizations, entity relationships, and more.
```

> When the xref:diagrams-extension-option.adoc[] is specifically enabled via the xref:options.adoc[], the {project-name} converts diagrams from text to image using a local or remote Kroki diagram server, and shows them as if they were local or remote image files in the HTML.

Reviewer: ```suggestion
When the xref:diagrams-extension-option.adoc[] is enabled via the xref:options.adoc[], the {project-name} converts diagrams from text to image using a local or remote Kroki diagram server, and shows them as if they were local or remote image files in the HTML.
```

Learn more on the xref:diagrams-extension-option.adoc[] page.

== Markdown friendly

AsciiDoc accommodates simple markdown in your AsciiDoc document.

This may save you time:

* Writing simple notes

> * Converting files from markdown to AsciiDoc gradually.

Reviewer: ```suggestion
* Converting files from Markdown to AsciiDoc gradually.
```

= Features
:url-asciidoctor-sc-image: https://cdn.jsdelivr.net/gh/asciidoctor/asciidoctor/screenshot.png

The main feature is viewing AsciiDoc as HTML5 directly in the browser.
Other features cover where those AsciiDoc files are located, and how they are rendered.

[#example]
== View AsciiDoc as HTML5

The screen capture below shows raw AsciiDoc on the left, and rendered HTML5 on the right.

If the extension is inactive, you'll see the raw AsciiDoc.
If it is active, you'll see the rendered HTML5.

Reproduce this example yourself by following the instructions further below in <<try-it-yourself>>.

image::{url-asciidoctor-sc-image}["Screen capture of two screens, showing raw AsciiDoc and rendered as HTML5"]

> 

Author: Add auto reload of changed content as a feature.

[#try-it-yourself]
=== Try it yourself

> 

Author: Move try it yourself up to level 2 from level 3

+
.When viewing AsciiDoc files often
TIP: You may change your computer settings to launch your web browser for `.adoc` files.
. Toggle the {project-name} activation by clicking its extension icon in the browser bar.

.Test file
[,asciidoc]
-----
= AsciiDoc is Writing Zen
Doc Writer <doc.writer@example.com>
:icons: font

_Zen_ in the *art* of writing `plain text` with
https://asciidoc.org[AsciiDoc]

[TIP]
Use https://asciidoctor.org[Asciidoctor] for the best AsciiDoc experience.footnote:[Not to mention the best looking output!]
Then icon:messages[role=aqua] about it!

== Sample section

[square]
* item 1
* item 2

[,ruby]
----
puts "Hello, World!"
----
-----

== Insert AsciiDoc attributes

Insert a set of attributes to each page viewed.
See xref:custom-attributes.adoc[].

== Change the presentation style

AsciiDoc separates content from presentation.

Change the stylesheet or CSS that is used to render HTML5 from the AsciiDoc.
For example, you may prefer a darker or a more compact presentation style.
See xref:options.adoc[].

== View AsciiDoc files from different locations

The extension allows you to view AsciiDoc files that are:

* Local to your browser (`file://`)
* Remote

> * Viewed in raw mode on hosted git providers including:

Author: Sounds to esoteric. Present in plainer language.
